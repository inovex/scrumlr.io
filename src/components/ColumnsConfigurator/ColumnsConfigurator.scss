@import "constants/style";

.columns-configurator {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-shrink: 0;
  flex-grow: 1;

  // allow horizontal scroll
  overflow: auto hidden;
  width: 100%;
  max-width: 100%;
  height: 100%;

  @include scrollbar();
}

.columns-configurator__columns-wrapper {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-shrink: 0;
  flex-grow: 1;
}
