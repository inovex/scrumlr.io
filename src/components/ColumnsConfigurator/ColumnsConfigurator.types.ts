export type ColumnPlacement = "first" | "center" | "last" | "all";

export type AddTemplateColumnAlignment = "left" | "right";
