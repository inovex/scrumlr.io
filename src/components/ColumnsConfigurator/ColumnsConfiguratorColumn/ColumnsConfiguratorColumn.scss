@import "constants/style";

$column-width: 158px;
$column-height: 340px;

.columns-configurator-column {
  display: flex;
  flex-direction: column;
  gap: $spacing--md;
  flex-shrink: 0;
  flex-grow: 1;

  width: $column-width;
  height: $column-height;

  padding: $spacing--md $spacing--base;

  border: $spacing--xxs solid var(--accent-color--light);

  z-index: $base-z-index;

  &--even {
    background-color: $gray--200;
  }
  &--odd {
    background-color: $gray--100;
  }

  // not the column that's currently being dragged, but actually where it was
  &--active-drag {
    background-color: rgba(var(--accent-color--100-rgb, 0.5));
    border: $spacing--xxs dashed var(--accent-color--light);
    z-index: $base-z-index + $base-z-index-step; // display dragged item above others
  }

  &--active-drop {
    background-color: rgba(var(--accent-color--100-rgb, 0.5));
    border: $spacing--xxs dashed var(--accent-color--light);
  }

  &--hidden {
    background-image: $column-stripes--light;
  }

  // deactivate borders depending on position
  &--border {
    &-first {
      border-right: none;
    }
    &-center {
      border-left: none;
      border-right: none;
    }
    &-last {
      border-left: none;
    }
  }
}

.columns-configurator-column__menu {
  position: relative;
  display: flex;
  flex-direction: column;
  gap: $spacing--xl; // TODO proper gaps
  align-items: center;
}

.columns-configurator-column__icon {
  width: $icon--large;
  height: $icon--large;

  &--dnd {
    width: 22px;
    height: 15px;
  }
}

.columns-configurator-column__drag-element {
  cursor: grab;
  touch-action: none; // prevent funky mobile dnd behaviour

  &--dragging {
    cursor: grabbing;
  }
}

// note: hardcoded values are always subject to change with the column or color picker themselves
.columns-configurator-column__color-picker {
  &--open {
    // only apply to opened color picker, as it behaves differently
    position: absolute;
    // some precise but hardcoded positioning values so it fits exactly on top of the active color
    // varying by column position (left, center, right) probably because of borders ig
    top: 40px - 1px;
    left: 40px + 1px;
  }

  // different column layout require different offsets ... would be nice to have a better solution for this in the future
  &--column-center {
    left: 40px + 3px;
  }

  &--column-all {
    left: 40px - 1px;
  }
}

.columns-configurator-column__button {
  all: unset;
  cursor: pointer;
  max-width: $icon--large;
  max-height: $icon--large;

  &:disabled {
    color: $gray--800;
    cursor: default;
  }
}

[theme="dark"] {
  .columns-configurator-column {
    border-color: var(--accent-color--dark);

    &--even {
      background-color: $navy--600;
    }
    &--odd {
      background-color: $navy--400;
    }

    &--active-drag {
      background-color: rgba(var(--accent-color--200-rgb, 0.5));
      border-color: var(--accent-color--dark);
    }

    &--active-drop {
      background-color: rgba(var(--accent-color--200-rgb, 0.5));
      border-color: var(--accent-color--dark);
    }

    &--hidden {
      background-image: $column-stripes--dark;
    }
  }

  .columns-configurator-column__icon {
    color: $gray--000;
  }
}
