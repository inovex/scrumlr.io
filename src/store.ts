import { createStore, compose } from 'redux';
import { reactReduxFirebase } from 'react-redux-firebase';

import rootReducer from './reducers';
import { firebase as fbConfig } from './config';
import * as firebase from 'firebase';

export default function configureStore(initialState: {}, history?: {}) {
  firebase.initializeApp(fbConfig);

  const composeEnhancers =
    (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const createStoreWithMiddleware = composeEnhancers(
    reactReduxFirebase(firebase, {
      userProfile: 'users',
      enableLogging: false
    })
  )(createStore);
  const store = createStoreWithMiddleware(rootReducer);

  // tslint:disable
  if ((module as any).hot) {
    // Enable Webpack hot module replacement for reducers
    (module as any).hot.accept('./reducers', () => {
      const nextRootReducer = require('./reducers');
      store.replaceReducer(nextRootReducer);
    });
  }
  // tslint:enable

  return store;
}
