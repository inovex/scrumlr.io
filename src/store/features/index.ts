export * from "./auth";
export * from "./board";
export * from "./boardReactions";
export * from "./columns";
export * from "./notes";
export * from "./participants";
export * from "./reactions";
export * from "./requests";
export * from "./skinTone";
export * from "./view";
export * from "./votes";
export * from "./votings";
export * from "./templates";
export * from "./templateColumns";
