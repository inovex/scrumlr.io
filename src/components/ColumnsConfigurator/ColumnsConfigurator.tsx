import {uniqueId} from "underscore";
import classNames from "classnames";
import {useState} from "react";
import {EditableTemplateColumn, TemplateColumn} from "store/features";
import {Color, getColorClassName, getNextColor, getPreviousColor} from "constants/colors";
import {closestCenter, DndContext, DragEndEvent, DragOverEvent, DragOverlay, DragStartEvent, PointerSensor, useSensor, useSensors} from "@dnd-kit/core";
import {horizontalListSortingStrategy, SortableContext} from "@dnd-kit/sortable";
import {ColumnsConfiguratorColumn} from "./ColumnsConfiguratorColumn/ColumnsConfiguratorColumn";
import {AddTemplateColumn} from "./AddTemplateColumn/AddTemplateColumn";
import {calcPlacement} from "./ColumnsConfigurator.utils";
import {AddTemplateColumnAlignment} from "./ColumnsConfigurator.types";
import "./ColumnsConfigurator.scss";

export type ColumnsConfiguratorProps = {
  className: string;
  templateId: string;
  columns: EditableTemplateColumn[]; // without deleted cols
  addColumn: (templateColumn: TemplateColumn, index: number) => void;
  moveColumn: (fromIndex: number, toIndex: number) => void;
  editColumn: (templateColumn: EditableTemplateColumn, overwrite: Partial<EditableTemplateColumn>) => void;
  deleteColumn: (templateColumn: EditableTemplateColumn) => void;
};

export const ColumnsConfigurator = (props: ColumnsConfiguratorProps) => {
  // id of column which is actively being dragged
  const [activeElementId, setActiveElementId] = useState<string | null>(null);
  const [activeColumn, setActiveColumn] = useState<EditableTemplateColumn | null>(null);
  const [dropElementId, setDropElementId] = useState<string | null>(null);

  const sensors = useSensors(useSensor(PointerSensor));

  const handleDragStart = (event: DragStartEvent) => {
    const activeId = event.active.id as string;
    setActiveElementId(activeId);
    setActiveColumn(props.columns.find((c) => c.id === activeId)!);
  };

  const handleDragOver = (event: DragOverEvent) => {
    setDropElementId((event.over?.id as string) ?? null);
  };

  const handleDragEnd = (event: DragEndEvent) => {
    const {active, over} = event;

    if (active.id !== over?.id) {
      const oldIndex = props.columns.findIndex((c) => c.id === active.id);
      const newIndex = props.columns.findIndex((c) => c.id === over?.id);

      props.moveColumn(oldIndex, newIndex);
    }

    // don't forget to clear states
    setActiveElementId(null);
    setActiveColumn(null);
    setDropElementId(null);
  };

  // function to calculate where each column would be positioned if the drag-and-drop operation were completed at its current state
  const getPotentialIndex = (index: number) => {
    // no valid drag operation occurring
    if (!activeElementId || !dropElementId) return index;

    const dragElementIndex = props.columns.findIndex((c) => c.id === activeElementId);
    const dropElementIndex = props.columns.findIndex((c) => c.id === dropElementId);

    // if current index matches active index of drag element, the column is being dragged there,
    // so the potential index is that of the drop element
    if (index === dragElementIndex) {
      return dropElementIndex;
    }

    // dragging right: active index less the drop zone, and element is between them, the element shifts to the left
    if (dragElementIndex < dropElementIndex && index > dragElementIndex && index <= dropElementIndex) {
      return index - 1;
    }

    // dragging left: analogue to above, element shifts to the right
    if (dragElementIndex > dropElementIndex && index < dragElementIndex && index >= dropElementIndex) {
      return index + 1;
    }

    // otherwise the index remains unchanged
    return index;
  };

  // renders the drag overlay, which is the element which is dragged
  const renderDragOverlay = () => {
    if (!activeElementId || !activeColumn) return null;
    const activeColumnIndex = getPotentialIndex(props.columns.findIndex((c) => c.id === activeColumn.id));

    return (
      <ColumnsConfiguratorColumn
        className={classNames("columns-configurator__column", "columns-configurator__column--ghost")}
        column={activeColumn}
        // here, use actual index unlike potential index in the main part
        index={activeColumnIndex}
        placement={calcPlacement(activeColumnIndex, props.columns)}
        allColumns={props.columns}
      />
    );
  };

  const addTemplateColumn = (alignment: AddTemplateColumnAlignment, color: Color) => {
    const newColumn: TemplateColumn = {
      id: uniqueId("col"),
      template: props.templateId,
      name: `Column ${props.columns.length + 1}`,
      description: "",
      color,
      visible: true,
      index: -1, // will be overwritten anyway
    };

    // target index for the right has to take deleted cols which are still part of the array into account,
    // which is why we use the total col amount as new index
    const index = alignment === "left" ? 0 : props.columns.length;

    props.addColumn(newColumn, index);
  };

  if (props.columns.length === 0) {
    return null;
  }

  return (
    <DndContext sensors={sensors} collisionDetection={closestCenter} onDragStart={handleDragStart} onDragOver={handleDragOver} onDragEnd={handleDragEnd}>
      <SortableContext items={props.columns} strategy={horizontalListSortingStrategy}>
        <div className={classNames(props.className, "columns-configurator", getColorClassName("backlog-blue"))}>
          <AddTemplateColumn alignment="left" color={getPreviousColor(props.columns[0].color)} onClick={addTemplateColumn} />
          <div className="columns-configurator__columns-wrapper">
            {props.columns.map((column, index) => {
              const potentialIndex = getPotentialIndex(index);
              return (
                <ColumnsConfiguratorColumn
                  className="columns-configurator__column"
                  key={column.id}
                  column={column}
                  index={potentialIndex}
                  activeDrag={column.id === activeElementId}
                  activeDrop={column.id === dropElementId}
                  placement={calcPlacement(potentialIndex, props.columns)}
                  allColumns={props.columns}
                  editColumn={props.editColumn}
                  deleteColumn={props.deleteColumn}
                />
              );
            })}
          </div>
          <AddTemplateColumn alignment="right" color={getNextColor(props.columns[props.columns.length - 1].color)} onClick={addTemplateColumn} />
        </div>
      </SortableContext>

      <DragOverlay>{renderDragOverlay()}</DragOverlay>
    </DndContext>
  );
};
