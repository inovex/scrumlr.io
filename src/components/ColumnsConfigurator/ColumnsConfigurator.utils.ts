import {EditableTemplateColumn, ExistenceStatus, Template, TemplateColumn, TemplateWithColumns} from "store/features";
import {ColumnPlacement} from "./ColumnsConfigurator.types";

export const calcPlacement = (index: number, columns: TemplateColumn[]): ColumnPlacement => {
  const {length} = columns;
  if (length === 1) {
    return "all";
  }
  if (index === 0) {
    return "first";
  }
  if (index === length - 1) {
    return "last";
  }
  return "center";
};

// convert TemplateColumn to EditableColumn, used for test purposes
export const convertToEditableColumn = (c: TemplateColumn, overrideExistenceStatus?: Partial<ExistenceStatus>): EditableTemplateColumn =>
  ({
    ...c,
    persisted: false,
    mode: undefined,
    ...overrideExistenceStatus,
  }) as EditableTemplateColumn;

// from an array of all templates and column, create a TemplateAndColumns array with all corresponding pairs
export const mergeTemplateAndColumns = (templates: Template[], templateColumns: TemplateColumn[]) =>
  templates.map((template) => ({
    template,
    columns: templateColumns.filter((column) => column.template === template.id),
  })) as TemplateWithColumns[];

export const getTemplateAndColumnsByTemplateId = (templatesAndColumns: {templates: Template[]; templateColumns: TemplateColumn[]}, templateId: string) =>
  mergeTemplateAndColumns(templatesAndColumns.templates, templatesAndColumns.templateColumns).find((tac) => tac.template.id === templateId);
