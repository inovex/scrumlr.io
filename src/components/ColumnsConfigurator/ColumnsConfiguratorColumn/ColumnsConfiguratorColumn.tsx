import classNames from "classnames";
import {EditableTemplateColumn} from "store/features";
import {ReactComponent as VisibleIcon} from "assets/icons/visible.svg";
import {ReactComponent as HiddenIcon} from "assets/icons/hidden.svg";
import {ReactComponent as DeleteIcon} from "assets/icons/trash.svg";
import {ReactComponent as DnDIcon} from "assets/icons/drag-and-drop.svg";
import {ColorPicker} from "components/ColorPicker/ColorPicker";
import {Color, COLOR_ORDER, getColorClassName} from "constants/colors";
import {useSortable} from "@dnd-kit/sortable";
import {CSSProperties, useCallback, useEffect, useState} from "react";
import {CSS} from "@dnd-kit/utilities";
import {useStripeOffset} from "utils/hooks/useStripeOffset";
import {ColumnConfiguratorColumnNameDetails, OpenState} from "./ColumnConfiguratorColumnNameDetails/ColumnConfiguratorColumnNameDetails";
import {ColumnPlacement} from "../ColumnsConfigurator.types";
import "./ColumnsConfiguratorColumn.scss";

export type ColumnsConfiguratorColumnProps = {
  className?: string;
  column: EditableTemplateColumn;
  index: number;
  placement?: ColumnPlacement;
  activeDrag?: boolean;
  activeDrop?: boolean;
  allColumns: EditableTemplateColumn[];
  editColumn?: (templateColumn: EditableTemplateColumn, overwrite: Partial<EditableTemplateColumn>) => void;
  deleteColumn?: (templateColumn: EditableTemplateColumn) => void;
};

export const ColumnsConfiguratorColumn = (props: ColumnsConfiguratorColumnProps) => {
  const [titleEditState, setTitleEditState] = useState<OpenState>("closed");
  const [openColorPicker, setOpenColorPicker] = useState(false);

  const {attributes, listeners, setNodeRef, transform, transition} = useSortable({id: props.column.id});
  const {
    bindings: {ref: stripeRef, style: stripeStyle},
    updateOffset,
  } = useStripeOffset<HTMLDivElement>();

  // combine ref to be used by both offset hook and dnd kit
  const combinedRef = useCallback(
    (node: HTMLDivElement | null) => {
      stripeRef.current = node;
      setNodeRef(node);
    },
    [stripeRef, setNodeRef]
  );

  // combine styles from both offset hook and dnd kit
  const style: CSSProperties = {
    ...stripeStyle,
    transform: CSS.Transform.toString(transform),
    transition,
  };

  // update offset when dragging or columns change
  useEffect(() => {
    updateOffset();
  }, [props.activeDrag, props.activeDrop, props.allColumns, updateOffset]);

  const editColor = (color: Color) => {
    props.editColumn?.(props.column, {color});
    setOpenColorPicker(false);
  };

  // disable delete if only one column remains
  const disableDelete = props.allColumns.length <= 1;

  return (
    <div
      className={classNames(
        props.className,
        "columns-configurator-column",
        `columns-configurator-column--border-${props.placement ?? "ghost"}`,
        {
          "columns-configurator-column--even": props.index % 2 === 0,
          "columns-configurator-column--odd": props.index % 2 !== 0,
          "columns-configurator-column--hidden": !props.column.visible,
          "columns-configurator-column--active-drag": props.activeDrag,
          "columns-configurator-column--active-drop": props.activeDrop,
        },
        getColorClassName(props.column.color)
      )}
      ref={combinedRef}
      style={style}
      {...attributes}
    >
      <ColumnConfiguratorColumnNameDetails
        className="columns-configurator-column__name-details"
        name={props.column.name}
        description={props.column.description}
        openState={titleEditState}
        setOpenState={setTitleEditState}
        updateColumnTitle={(name, description) => props.editColumn?.(props.column, {name, description})}
      />
      {titleEditState === "closed" && (
        <div className="columns-configurator-column__menu">
          <DnDIcon
            className={classNames("columns-configurator-column__icon", "columns-configurator-column__icon--dnd", "columns-configurator-column__drag-element", {
              "columns-configurator-column__drag-element--dragging": props.activeDrag,
            })}
            {...listeners}
            data-cy="columns-configurator-column__drag-element"
          />
          <ColorPicker
            className={classNames(
              "columns-configurator-column__color-picker",
              {"columns-configurator-column__color-picker--open": openColorPicker},
              `columns-configurator-column__color-picker--column-${props.placement}`
            )}
            open={openColorPicker}
            colors={COLOR_ORDER}
            activeColor={props.column.color}
            selectColor={editColor}
            attemptOpenColorPicker={() => setOpenColorPicker(true)} // always allow to open
            closeColorPicker={() => setOpenColorPicker(false)}
            fitted
            dataCy="columns-configurator-column__color-picker"
          />
          <button
            className={classNames("columns-configurator-column__button", {"columns-configurator-column__button--disabled": false}, "column-configurator-column__visibility-button")}
            onClick={() => props.editColumn?.(props.column, {visible: !props.column.visible})}
          >
            {props.column.visible ? (
              <VisibleIcon
                className={classNames("columns-configurator-column__icon", "columns-configurator-column__icon--visible")}
                data-cy="columns-configurator-column__icon--visibility"
              />
            ) : (
              <HiddenIcon
                className={classNames("columns-configurator-column__icon", "columns-configurator-column__icon--hidden")}
                data-cy="columns-configurator-column__icon--visibility"
              />
            )}
          </button>
          <button
            className={classNames(
              "columns-configurator-column__button",
              {"columns-configurator-column__button--disabled": disableDelete},
              "column-configurator-column__delete-button"
            )}
            onClick={() => props.deleteColumn?.(props.column)}
            disabled={disableDelete}
          >
            <DeleteIcon className={classNames("columns-configurator-column__icon", "columns-configurator-column__icon--delete")} />
          </button>
        </div>
      )}
    </div>
  );
};
