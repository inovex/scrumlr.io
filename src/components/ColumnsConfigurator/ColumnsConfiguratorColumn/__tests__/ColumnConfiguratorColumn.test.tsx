import {render} from "testUtils";
import {Provider} from "react-redux";
import getTestStore from "utils/test/getTestStore";
import {ColumnsConfiguratorColumn} from "../ColumnsConfiguratorColumn";
import {EditableTemplateColumn} from "store/features";
import {MockStoreEnhanced} from "redux-mock-store";
import {ApplicationState} from "store";
import {fireEvent} from "@testing-library/react";
import {convertToEditableColumn} from "../../ColumnsConfigurator.utils";

describe("ColumnConfiguratorColumn render", () => {
  let testStore: MockStoreEnhanced<ApplicationState>;

  beforeEach(() => {
    testStore = getTestStore();
  });

  // different way of rendering the columns configurator column
  /* const renderColumnConfiguratorColumn = (override: Partial<ColumnsConfiguratorColumnProps> = {}) => {
    const testTemplatesColumns = getTestApplicationState().templatesColumns
    const testEditableTemplatesColumns = testTemplatesColumns.map(c => convertToEditableColumn(c))
    const editableTemplatesColumn0 = testEditableTemplatesColumns[0]
    const editableTemplatesColumn1 = testEditableTemplatesColumns[1]
    const allColumns = [editableTemplatesColumn0, editableTemplatesColumn1]

    const defaultProps: ColumnsConfiguratorColumnProps = {
      className: "columns-configurator-column",
      column: editableTemplatesColumn0,
      index: 0,
      allColumns,
      editColumn: jest.fn(),
    };

    return render(<ColumnsConfiguratorColumn {...defaultProps} {...override}/>);
  }; */

  const renderColumnConfiguratorColumn = (
    templateId: string,
    columnId: string,
    potentialIndex: number,
    editColumn?: (templateColumn: EditableTemplateColumn, overwrite: Partial<EditableTemplateColumn>) => void,
    deleteColumn?: (templateColumn: EditableTemplateColumn) => void
  ) => {
    // some preprocessing to get the correct column and convert to editableColumn, which would normally be done by TemplateEditor component
    const columns = testStore.getState().templateColumns.filter((tc) => tc.template === templateId);
    const editableColumns = columns.map((bc) => convertToEditableColumn(bc));
    const column = editableColumns.find((c) => c.id === columnId)!;

    return render(
      <Provider store={testStore}>
        <ColumnsConfiguratorColumn column={column} index={potentialIndex} allColumns={editableColumns} editColumn={editColumn} deleteColumn={deleteColumn} />
      </Provider>
    );
  };

  it("should render correctly", () => {
    const {container} = renderColumnConfiguratorColumn("test-templates-id-1", "test-templates-columns-id-1", 0);

    expect(container).toMatchSnapshot();
  });

  it("should be an even column with index 0", () => {
    const {container} = renderColumnConfiguratorColumn("test-templates-id-1", "test-templates-columns-id-1", 0);
    const rootNode = container.firstChild!;

    expect(rootNode).toHaveClass("columns-configurator-column--even");
  });

  it("should be an odd column with index 1", () => {
    const {container} = renderColumnConfiguratorColumn("test-templates-id-1", "test-templates-columns-id-1", 1);
    const rootNode = container.firstChild!;

    expect(rootNode).toHaveClass("columns-configurator-column--odd");
  });

  // opens color picker, select pink color, then checks if editColumn fn has been called
  it("should be able to change color", async () => {
    const columnId = "test-templates-columns-id-1";
    const editColumnSpy: jest.Mock<(templateColumn: EditableTemplateColumn, overwrite: Partial<EditableTemplateColumn>) => void> = jest.fn();
    const {container} = renderColumnConfiguratorColumn("test-templates-id-1", columnId, 0, editColumnSpy);

    const colorPickerElement = container.querySelector(".columns-configurator-column__color-picker")!;
    expect(colorPickerElement).toBeInTheDocument();
    fireEvent.click(colorPickerElement);

    const colorPickerButtonPlanningPink = container.querySelector<HTMLButtonElement>("li.accent-color__planning-pink button")!;

    fireEvent.mouseDown(colorPickerButtonPlanningPink);

    expect(editColumnSpy).toHaveBeenCalledWith(expect.anything(), expect.objectContaining({color: "planning-pink"}));
  });

  it("should be able to toggle visibility", () => {
    const editColumnSpy: jest.Mock<(templateColumn: EditableTemplateColumn, overwrite: Partial<EditableTemplateColumn>) => void> = jest.fn();
    const {container} = renderColumnConfiguratorColumn("test-templates-id-1", "test-templates-columns-id-1", 0, editColumnSpy);
    const visibilityButtonElement = container.querySelector(".column-configurator-column__visibility-button")! as HTMLButtonElement;

    fireEvent.click(visibilityButtonElement);
    expect(editColumnSpy).toHaveBeenCalledWith(expect.anything(), expect.objectContaining({visible: false}));
  });

  it("should be deletable", () => {
    const deleteColumnSpy: jest.Mock<(templateColumn: EditableTemplateColumn) => void> = jest.fn();
    const {container} = renderColumnConfiguratorColumn("test-templates-id-1", "test-templates-columns-id-1", 0, undefined, deleteColumnSpy);
    const deleteButtonElement = container.querySelector(".column-configurator-column__delete-button")! as HTMLButtonElement;

    expect(deleteButtonElement).not.toBeDisabled();
    fireEvent.click(deleteButtonElement);
    expect(deleteColumnSpy).toHaveBeenCalledWith(expect.objectContaining({id: "test-templates-columns-id-1"}));
  });
});
