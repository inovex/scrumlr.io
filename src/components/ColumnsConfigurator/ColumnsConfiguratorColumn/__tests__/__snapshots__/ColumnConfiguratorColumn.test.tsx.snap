// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnConfiguratorColumn render should render correctly 1`] = `
<div>
  <div
    aria-describedby=""
    aria-disabled="false"
    aria-roledescription="sortable"
    class="columns-configurator-column columns-configurator-column--border-ghost columns-configurator-column--even accent-color__backlog-blue"
    role="button"
    style="--stripe-offset: 0px;"
    tabindex="0"
  >
    <div
      class="columns-configurator-column__name-details column-configurator-column-name-details__name-wrapper"
    >
      <input
        autocomplete="off"
        class="column-configurator-column-name-details__name"
        placeholder="Title"
        value="sample templates col name 1"
      />
      <div
        class="column-configurator-column-name-details__inline-description"
        role="button"
        tabindex="0"
      >
        sample templates col name 1
      </div>
    </div>
    <div
      class="columns-configurator-column__menu"
    >
      <svg
        class="columns-configurator-column__icon columns-configurator-column__icon--dnd columns-configurator-column__drag-element"
        data-cy="columns-configurator-column__drag-element"
      >
        drag-and-drop.svg
      </svg>
      <span
        aria-label="current colorBacklog Blue"
        class="columns-configurator-column__color-picker columns-configurator-column__color-picker--column-undefined color-picker__color-option color-picker__color-option--selected"
        data-cy="columns-configurator-column__color-picker"
        role="button"
        tabindex="0"
      />
      <button
        class="columns-configurator-column__button column-configurator-column__visibility-button"
      >
        <svg
          class="columns-configurator-column__icon columns-configurator-column__icon--visible"
          data-cy="columns-configurator-column__icon--visibility"
        >
          visible.svg
        </svg>
      </button>
      <button
        class="columns-configurator-column__button column-configurator-column__delete-button"
      >
        <svg
          class="columns-configurator-column__icon columns-configurator-column__icon--delete"
        >
          trash.svg
        </svg>
      </button>
    </div>
  </div>
</div>
`;
